src/Microsoft.AspNetCore.Http.Abstractions/Extensions/UseWhenExtensions.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    using Predicate = Func<HttpContext, bool>;
+
+    /// <summary>
+    /// Extension methods for <see cref="IApplicationBuilder"/>.
+    /// </summary>
+    public static class UseWhenExtensions
+    {
+        /// <summary>
+        /// Conditionally creates a branch in the request pipeline that is rejoined to the main pipeline.
+        /// </summary>
+        /// <param name="app"></param>
+        /// <param name="predicate">Invoked with the request environment to determine if the branch should be taken</param>
+        /// <param name="configuration">Configures a branch to take</param>
+        /// <returns></returns>
+        public static IApplicationBuilder UseWhen(this IApplicationBuilder app, Predicate predicate, Action<IApplicationBuilder> configuration)
```

Comment:
`if (condition) { } else { }` can be done as `if (condition) { } if (!condition) { }`, but it may be inefficient to evaluate the conditional twice.

`if (condition) { } elseif (condition) else { }` can be done as `if (condition) { } else { if (condition) { } else { }  }`